includes/backends/class-omise-backend-installment.php

```diff
@@ -0,0 +1,124 @@
+ * @method public get_available_providers
+ * @method public filter_available_terms
+ * @method public calculate_monthly_payment_amount
+ */
+class Omise_Backend_Installment extends Omise_Backend {
+	/**
+	 * @var array  of known installment providers.
+	 */
+	protected static $providers = array();
+
+	public function initiate() {
+		self::$providers = array(
+			'installment_first_choice' => array(
+				'bank_code'          => 'first_choice',
+				'title'              => __( 'Krungsri First Choice', 'omise' ),
+				'interest_rate'      => 0.013,
+				'min_allowed_amount' => 300.00,
+			),
+
+			'installment_bay' => array(
+				'bank_code'          => 'bay',
+				'title'              => __( 'Krungsri', 'omise' ),
+				'interest_rate'      => 0.008,
+				'min_allowed_amount' => 300.00,
+			),
+
+			'installment_ktc' => array(
+				'bank_code'          => 'ktc',
+				'title'              => __( 'Krungthai Card (KTC)', 'omise' ),
+				'interest_rate'      => 0.008,
+				'min_allowed_amount' => 300.00,
+			),
+
+			'installment_bbl' => array(
+				'bank_code'          => 'bbl',
+				'title'              => __( 'Bangkok Bank', 'omise' ),
+				'interest_rate'      => 0.008,
+				'min_allowed_amount' => 500.00,
+			),
+
+			'installment_kbank' => array(
+				'bank_code'          => 'kbank',
+				'title'              => __( 'Kasikorn Bank', 'omise' ),
+				'interest_rate'      => 0.0065,
+				'min_allowed_amount' => 500.00,
+			),
+		);
+	}
+
+	/**
+	 * @param  string $currency
+	 * @param  float  $purchase_amount
+	 *
+	 * @return array  of an available installment providers
+	 */
+	public function get_available_providers( $currency, $purchase_amount ) {
+		// Note: as installment payment at the moment only supports for THB currency, so the 
+		//       $purchase_amount is multiplied with 100 to convert the amount into subunit (satang).
+		$providers = $this->capabilities()->getInstallmentBackends( $currency, ( $purchase_amount * 100 ) );
+
+		foreach ( $providers as &$provider ) {
+			$provider->provider_code             = str_replace( 'installment_', '', $provider->_id );
+			$provider->provider_name             = isset( self::$providers[ $provider->_id ] ) ? self::$providers[ $provider->_id ]['title'] : strtoupper( $provider->code );
+			$provider->allowed_installment_terms = $this->filter_available_terms( $provider->allowed_installment_terms, self::$providers[ $provider->_id ], $purchase_amount );
+			$provider->interest                  = $this->capabilities()->is_zero_interest() ? 0 : ( self::$providers[ $provider->_id ]['interest_rate'] * 100 );
+		}
+
+		return $providers;
+	}
+
+	/**
+	 * @param  array $available_terms
+	 * @param  array $provider_detail
+	 * @param  float $purchase_amount
+	 *
+	 * @return array  of an filtered available terms
+	 */
+	public function filter_available_terms( $available_terms, $provider_detail, $purchase_amount ) {
+		$filtered_available_terms = array();
+
+		sort( $available_terms );
```

Comment:
Should a filter function be sorting?

Reply:
@jonrandy Hi, I didn't change this one at the recent commit as we were discussing about this misled-method name.
Let me know if you still have any further opinion about sorting here. : D

Reply:
I'm still not overly keen on `term` in the code referring to 2 different things - both a single number representing the number of months (which isn't what term means - a better name would be `terms` or `term_count`), and an associative array containing the number of terms and the monthly amount

Reply:
@jonrandy you mean something like these 2 cases 👇 ?

" This BBL has 3 installment terms you can choose, which are '3', '6', '9'. "

and

available_terms = "3", "6", "9"
term 3, term 6, term 9.

Reply:
The available terms are a term of 3 months, a term of 6 months, and a term of 9 months. 'Term' is the time period itself, not it's length.

Aside from the naming issue, my problem with the above function was the use of the word term to refer to both:

a) The length of the term over which the instalments are paid

and 

b) A construct representing the details of a particular instalment plan (the monthly payment amount and the term length)

This dual use of the same word seems wrong

Reply:
Uhmm... my English is error today, any suggestion? @jonrandy 😂 
Where it should be replaced with `term_length` or `term_months`?

Reply:
```php
$provider->allowed_installment_terms = $this->get_valid_terms(...);

public function get_valid_terms() {
    $valid_terms = array();

    return $valid_terms;
}
```
👆 These terms context seem ok to me. Maybe inside `foreach` one?

Reply:
A better name may even be `get_valid_instalment_plans`, given what is actually being returned

Reply:
An attempt at a rewrite... might not be 100% perfect.

```php
<?php
/**
 * Note: all calculations in this class are based only on Thailand VAT and fee
 *       as currently Installment feature supported only for merchants
 *       that have registered with Omise Thailand account.
 *
 * @since 3.4
 *
 * @method public initiate
 * @method public get_available_providers
 * @method public get_valid_plans
 * @method public calculate_monthly_payment_amount
 */
class Omise_Backend_Installment extends Omise_Backend {
	/**
	 * @var array  of known installment providers.
	 */
	protected static $providers = array();

	public function initiate() {
		self::$providers = array(
			'installment_first_choice' => array(
				'bank_code'          => 'first_choice',
				'title'              => __( 'Krungsri First Choice', 'omise' ),
				'interest_rate'      => 0.013,
				'min_allowed_amount' => 300.00,
			),

			'installment_bay' => array(
				'bank_code'          => 'bay',
				'title'              => __( 'Krungsri', 'omise' ),
				'interest_rate'      => 0.008,
				'min_allowed_amount' => 300.00,
			),

			'installment_ktc' => array(
				'bank_code'          => 'ktc',
				'title'              => __( 'Krungthai Card (KTC)', 'omise' ),
				'interest_rate'      => 0.008,
				'min_allowed_amount' => 300.00,
			),

			'installment_bbl' => array(
				'bank_code'          => 'bbl',
				'title'              => __( 'Bangkok Bank', 'omise' ),
				'interest_rate'      => 0.008,
				'min_allowed_amount' => 500.00,
			),

			'installment_kbank' => array(
				'bank_code'          => 'kbank',
				'title'              => __( 'Kasikorn Bank', 'omise' ),
				'interest_rate'      => 0.0065,
				'min_allowed_amount' => 500.00,
			),
		);
	}

	/**
	 * @param  string $currency
	 * @param  float  $purchase_amount
	 *
	 * @return array  of an available installment providers
	 */
	public function get_available_providers( $currency, $purchase_amount ) {
		// Note: as installment payment at the moment only supports for THB currency, so the 
		//       $purchase_amount is multiplied with 100 to convert the amount into subunit (satang).
		$providers = $this->capabilities()->getInstallmentBackends( $currency, ( $purchase_amount * 100 ) );

		foreach ( $providers as &$provider ) {
			$provider->provider_code             = str_replace( 'installment_', '', $provider->_id );
			$provider->provider_name             = isset( self::$providers[ $provider->_id ] ) ? self::$providers[ $provider->_id ]['title'] : strtoupper( $provider->code );
			$provider->allowed_installment_plans = $this->get_valid_plans( $provider->allowed_installment_terms, self::$providers[ $provider->_id ], $purchase_amount );
			$provider->interest_rate             = $this->capabilities()->is_zero_interest() ? 0 : ( self::$providers[ $provider->_id ]['interest_rate'] * 100 );
		}

		return $providers;
	}

	/**
	 * @param  array $available_term_lengths (months)
	 * @param  array $provider_detail
	 * @param  float $purchase_amount
	 *
	 * @return array  of available installment plans
	 */
	public function get_valid_plans( $available_term_lengths, $provider_detail, $purchase_amount ) {
		$valid_plans = array();

		sort( $available_term_lengths );

		foreach ( $available_term_lengths as $term_length ) {
			$monthly_amount = $this->calculate_monthly_payment_amount(
				$purchase_amount,
				$term_length,
				// *** Wasn't this is_zero_interest check done before? ***
				$this->capabilities()->is_zero_interest() ? 0 : $provider_detail['interest_rate']
			);

			if ( $monthly_amount < $provider_detail['min_allowed_amount'] ) {
				break;
			}

			$valid_plans[] = array(
				'term_length'           => $term_length,
				'monthly_amount' 				=> $monthly_amount
			);
		}

		return $valid_plans;
	}

	/**
	 * @param  float $purchase_amount
	 * @param  int   $term             installment term.
	 * @param  float $interest_rate    its value can be '0' if merchant absorbs an interest.
	 *
	 * @return float  of a installment monthly payment (round up to 2 decimals).
	 */
	public function calculate_monthly_payment_amount( $purchase_amount, $term, $interest_rate ) {
		$interest = $purchase_amount * $term_length * $interest_rate;
		return round( ( $purchase_amount + $interest ) / $term_length, 2 );
	}
}
```

Reply:
File: `includes/backends/class-omise-backend-installment.php`
```php
$provider->available_installment_plans = $this-> get_available_instalment_plans(...);

public function get_available_instalment_plans( $allowed_terms, $provider_detail, $purchase_amount ) {
    $plans = array();

    foreach ( $allowed_terms as $term ) {
        $monthly_amount = $this->calculate_monthly_payment_amount( ... );

        $plans[] = array(
            'term'           => $term,
            'monthly_amount' => $monthly_amount
        );
    }

    return $plans;
}
```

File: `templates/payment/form-installment.php`
```html
<select ...>
    <option>Select term</option>
    <?php foreach ( $backend->available_installment_plans as $installment_plan ) : ?>
        <option value="<?php echo $installment_plan['term']; ?>">
            <?php
            echo sprintf(
                __( '%d months', 'omise', 'omise_installment_term_option' ),
                $installment_plan['term']
            );
            ?>

            <?php
            echo sprintf(
                __( '( %s / months )', 'omise', 'omise_installment_payment_per_month' ),
                wc_price( $installment_plan['monthly_amount'] )
            );
            ?>
        </option>
    <?php endforeach; ?>
</select>
```

@jonrandy LGTY?

Reply:
'term_length' instead of 'term'

Reply:
@jonrandy 👍 thanks for the input. Checked your rewrite-code. 
Any idea why `interest_rate` is better name than `interest` here?
```php
foreach ( $providers as &$provider ) {
	$provider->provider_code             = str_replace( 'installment_', '', $provider->_id );
	$provider->provider_name             = isset( self::$providers[ $provider->_id ] ) ? self::$providers[ $provider->_id ]['title'] : strtoupper( $provider->code );
	$provider->allowed_installment_plans = $this->get_valid_plans( $provider->allowed_installment_terms, self::$providers[ $provider->_id ], $purchase_amount );
	$provider->interest_rate             = $this->capabilities()->is_zero_interest() ? 0 : ( self::$providers[ $provider->_id ]['interest_rate'] * 100 );
}
```

Reply:
Because `interest` and `interest_rate` are entirely different things